pkg/autoscaler/statforwarder/forwarder.go

```diff
@@ -93,49 +87,70 @@ type Forwarder struct {
 	// on when shutting down.
 	processingWg sync.WaitGroup
 
-	// id2ip stores the IP extracted from the holder identity to avoid
-	// string split each time.
-	id2ip map[string]string
-
 	statCh chan stat
 	stopCh chan struct{}
 }
 
 // New creates a new Forwarder.
 func New(ctx context.Context, logger *zap.SugaredLogger, kc kubernetes.Interface, selfIP string, bs *hash.BucketSet, accept statProcessor) *Forwarder {
-	ns := system.Namespace()
 	bkts := bs.Buckets()
 	endpointsInformer := endpointsinformer.Get(ctx)
 	f := &Forwarder{
-		selfIP:          selfIP,
+		logger:     logger,
+		bs:         bs,
+		processors: make(map[string]bucketProcessor, len(bkts)),
+		statCh:     make(chan stat, 1000),
+		stopCh:     make(chan struct{}),
+	}
+
+	f.processingWg.Add(1)
+	go f.process()
+
+	lt := &leaseTracker{
 		logger:          logger,
+		selfIP:          selfIP,
+		bs:              bs,
 		kc:              kc,
 		endpointsLister: endpointsInformer.Lister(),
-		bs:              bs,
-		processors:      make(map[string]bucketProcessor, len(bkts)),
-		accept:          accept,
 		id2ip:           make(map[string]string),
-		statCh:          make(chan stat, 1000),
-		stopCh:          make(chan struct{}),
+		accept:          accept,
+		fwd:             f,
 	}
 
-	f.processingWg.Add(1)
-	go f.process()
-
 	leaseInformer := leaseinformer.Get(ctx)
 	leaseInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
-		FilterFunc: f.filterFunc(ns),
+		FilterFunc: lt.filterFunc(system.Namespace()),
 		Handler: cache.ResourceEventHandlerFuncs{
-			AddFunc:    f.leaseUpdated,
-			UpdateFunc: controller.PassNew(f.leaseUpdated),
+			AddFunc:    lt.leaseUpdated,
+			UpdateFunc: controller.PassNew(lt.leaseUpdated),
 			// TODO(yanweiguo): Set up DeleteFunc.
 		},
 	})
 
 	return f
 }
 
-func (f *Forwarder) extractIP(holder string) (string, error) {
+type leaseTracker struct {
```

Comment:
some comments about relationship to the other objects?

Reply:
done